crates/ursa-gateway/src/worker/cache/server.rs

```diff
@@ -2,44 +2,41 @@ use std::sync::Arc;
 
 use anyhow::anyhow;
 use async_trait::async_trait;
+use axum::response::IntoResponse;
 use axum::{
-    body::{HttpBody, StreamBody},
+    body::HttpBody,
     http::{response::Parts, StatusCode},
 };
 use bytes::BufMut;
+use hyper::Body;
 use tokio::{
     io::{duplex, AsyncWriteExt, DuplexStream},
     spawn,
     sync::{mpsc::UnboundedSender, oneshot},
 };
 use tokio_util::io::ReaderStream;
-use tracing::{error, info_span, Instrument, Span};
+use tracing::{error, info, info_span, Instrument, Span};
 use tracing_opentelemetry::OpenTelemetrySpanExt;
 
 use super::{Cache, CacheCommand};
 use crate::util::error::Error;
 
 #[async_trait]
 pub trait ServerCache: Send + Sync + 'static {
-    async fn get_announce(
-        &self,
-        k: &str,
-        no_cache: bool,
-    ) -> Result<StreamBody<ReaderStream<DuplexStream>>, Error>;
+    async fn get_announce(&self, k: &str, no_cache: bool) -> Result<StreamBody, Error>;
 }
 
 #[async_trait]
 impl ServerCache for Cache {
-    async fn get_announce(
-        &self,
-        k: &str,
-        no_cache: bool,
-    ) -> Result<StreamBody<ReaderStream<DuplexStream>>, Error> {
-        let (mut w, r) = duplex(self.stream_buf as usize);
+    async fn get_announce(&self, k: &str, no_cache: bool) -> Result<StreamBody, Error> {
         if no_cache {
             let span = info_span!("Cache invalidate");
-            fetch_and_insert(k, &self.tx, w).instrument(span).await?;
+            fetch(k, &self.tx)
```

Comment:
Essentially, `no cache` or `cache miss` are the same, right?

Reply:
I didn't noticed that the argument was coming from the HTTP cache control header. Will update.